File: Language/C#/Book/tmp/LINQ实战/Chapter11.CSharp/Chapter11.WinForms/Form1.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using System.Linq;
using System.Xml.Linq;
using System.Data.Linq;
using AmazonCommon;
using LinqInAction.LinqToSql;

namespace Chapter11.WinForms {
  public partial class ImportForm : Form {

		LinqInActionDataContext ctx;
    XNamespace ns = Amazon.AmazonNS;
    XElement amazonXml;

    public ImportForm() {
      InitializeComponent();
      this.Load += new EventHandler(ImportForm_Load);
			ctx = new LinqInActionDataContext();
    }

    void ImportForm_Load(object sender, EventArgs e) {
      subjectComboBox.DataSource = ctx.Subjects.ToList();
    }

    void searchButton_Click(object sender, EventArgs e) {

      var signHelper = new SignedRequestHelper(Amazon.AccessKey, Amazon.SecretKey, "ecs.amazonaws.com");

      var parameters = new Dictionary<String, String>() {
        { "Service", Amazon.ServiceName},
        { "Version", Amazon.ServiceVersion },
        { "Operation", "ItemSearch" },
        { "SearchIndex", "Books" },
        { "ResponseGroup", "Large" },
        { "Keywords", keywords.Text } 
      };

      string requestUrl = signHelper.Sign(parameters);
      amazonXml = XElement.Load(requestUrl);

      var books =
         from result in amazonXml.Descendants(ns + "Item")
					let attributes = result.Element(ns + "ItemAttributes")
					select new Book {
							Isbn=(string) attributes.Element(ns + "ISBN"), 
							Title=(string) attributes.Element(ns + "Title"),
					};

      bookBindingSource.DataSource = books;
    
      var rows =
        from row in bookDataGridView.Rows.OfType<DataGridViewRow>()
				let dataBoundBook =((Book) row.DataBoundItem)
			  join book in ctx.Books on dataBoundBook.Isbn equals book.Isbn.Trim()
				select row;

      foreach(DataGridViewRow row in rows) {
        row.DefaultCellStyle.BackColor = Color.LightGray;
        row.Cells[0].ReadOnly = true;
        row.Cells[1].Value = "** Already Exists ** - " + row.Cells[1].Value;
      }
    }

    void importButton_Click(object sender, EventArgs e) {
      var selectedBooks =
        from row in bookDataGridView.Rows.OfType<DataGridViewRow>()
				where ((bool) row.Cells[0].EditedFormattedValue) == true
				select (Book) row.DataBoundItem;

			using (var newContext = new LinqInActionDataContext()) {
				var booksToImport =
					from amazonItem in amazonXml.Descendants(ns + "Item")
					join selectedBook in selectedBooks
						on (string)amazonItem.Element(ns + "ItemAttributes").Element(ns + "ISBN")
							equals selectedBook.Isbn
					join p in newContext.Publishers
						on ((string)amazonItem.Element(ns + "ItemAttributes").Element(ns + "Publisher")).ToUpper()
							equals p.Name.Trim().ToUpper() into publishers
					from existingPublisher in publishers.DefaultIfEmpty()
					let attributes = amazonItem.Element(ns + "ItemAttributes")
					select new Book {
						ID = Guid.NewGuid(),
						Isbn = (string)attributes.Element(ns + "ISBN"),
						Title = (string)attributes.Element(ns + "Title"),
						Publisher = (existingPublisher ??
											 new Publisher {
												 ID = Guid.NewGuid(),
												 Name = (string)attributes.Element(ns + "Publisher")
											 }
						),
						Subject = (Subject)subjectComboBox.SelectedItem,
						PubDate = (DateTime)attributes.Element(ns + "PublicationDate"),
						Price = ParsePrice(attributes.Element(ns + "ListPrice")),
						BookAuthors = GetAuthors(attributes.Elements(ns + "Author"))
					};

				newContext.Subjects.Attach((Subject)subjectComboBox.SelectedItem);
				foreach (Book book in booksToImport) {
					newContext.Books.InsertOnSubmit(book);
				}

				try {
					newContext.SubmitChanges();
					MessageBox.Show(booksToImport.Count() + " books imported.");
				}
				catch (Exception ex) {
					MessageBox.Show("An error occured while attempting to import the selected books. " + Environment.NewLine + Environment.NewLine + ex.Message);
				}
			}
    }

    private EntitySet<BookAuthor> GetAuthors(IEnumerable<XElement> authorElements) {
      var bookAuthors =
        from authorElement in authorElements
        join author in ctx.Authors on (string) authorElement equals author.FirstName + " " + author.LastName into authors
        from existingAuthor in authors.DefaultIfEmpty()
        let nameParts = ((string)authorElement).Split(' ')
        select new BookAuthor {
          ID=Guid.NewGuid(),
          Author=existingAuthor ?? 
            new Author {
              ID=Guid.NewGuid(),
              FirstName=nameParts[0], 
              LastName=nameParts[1]
            }
        };

      EntitySet<BookAuthor> set = new EntitySet<BookAuthor>();
      set.AddRange(bookAuthors);
      return set;
    }

    private decimal ParsePrice(XElement priceElement) {
      return Convert.ToDecimal(((string)priceElement.Element(ns + "FormattedPrice")).Replace("$", String.Empty));
    }
  }
}

File: Language/C#/Book/tmp/LINQ实战/Chapter11.VB/Chapter11.FilteringAndMixingRelationalAndXml/Form1.vb
Imports System.Xml.Linq
Imports System.Data.Linq
Imports LinqInAction.LinqToSql
Imports AmazonCommon

Public Class Form1
  Dim ns As XNamespace = Amazon.AmazonNS
  Dim amazonXml As XElement

  Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
    subjectComboBox.DataSource = New LinqInActionDataContext().Subjects.ToList()
  End Sub

  Private Sub searchButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles searchButton.Click
    Dim signHelper = New SignedRequestHelper(Amazon.AccessKey, Amazon.SecretKey, "ecs.amazonaws.com")

    Dim parameters = New Dictionary(Of String, String)()
    parameters("Service") = Amazon.ServiceName
    parameters("Version") = Amazon.ServiceVersion
    parameters("Operation") = "ItemSearch"
    parameters("SearchIndex") = "Books"
    parameters("ResponseGroup") = "Large"
    parameters("Keywords") = keywords.Text

    Dim requestUrl = signHelper.Sign(parameters)
    amazonXml = XElement.Load(requestUrl)

    Dim books = From result In amazonXml.Descendants(ns + "Item") _
    Let attributes = result.Element(ns + "ItemAttributes") _
    Select New Book With { _
    .Isbn = CType(attributes.Element(ns + "ISBN"), String), _
    .Title = CType(attributes.Element(ns + "Title"), String) _
    }

    bookBindingSource.DataSource = books

    Dim ctx As LinqInActionDataContext = New LinqInActionDataContext()
    Dim rows = From row In bookDataGridView.Rows.OfType(Of DataGridViewRow)() _
     Join book In ctx.Books On CType(row.DataBoundItem, Book).Isbn.Trim() Equals book.Isbn.Trim() _
     Select row

    Console.WriteLine(rows.Count())

    For Each row As DataGridViewRow In rows
      row.DefaultCellStyle.BackColor = Color.LightGray
      row.Cells(0).ReadOnly = True
      row.Cells(1).Value = "** Already Exists ** - " + row.Cells(1).Value
    Next
  End Sub

  Private Sub importButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles importButton.Click
    Dim selectedBooks = From row In bookDataGridView.Rows.OfType(Of DataGridViewRow)() _
     Where CType(row.Cells(0).EditedFormattedValue, Boolean) = True _
     Select CType(row.DataBoundItem, Book)

    Dim ctx As LinqInActionDataContext = New LinqInActionDataContext()

    Dim booksToImport = From amazonItem In amazonXml.Descendants(ns + "Item") _
     Join selectedBook In selectedBooks On CType(amazonItem.Element(ns + "ItemAttributes").Element(ns + "ISBN"), String).Trim() Equals selectedBook.Isbn.Trim() _
     Group Join publisher In ctx.Publishers On CType(amazonItem.Element(ns + "ItemAttributes").Element(ns + "Publisher"), String).Trim() Equals publisher.Name.Trim() Into Group _
     From existingPublisher In Group.DefaultIfEmpty() _
     Let attributes = amazonItem.Element(ns + "ItemAttributes") _
     Select New Book With { _
     .ID = Guid.NewGuid(), _
     .Isbn = CType(attributes.Element(ns + "ISBN"), String), _
     .Title = CType(attributes.Element(ns + "Title"), String), _
     .Subject = CType(subjectComboBox.SelectedItem, Subject), _
     .PubDate = CType(attributes.Element(ns + "PublicationDate"), Date), _
     .Price = ParsePrice(attributes.Element(ns + "ListPrice")), _
     .Publisher = If(existingPublisher IsNot Nothing, existingPublisher, New Publisher With { _
     .ID = Guid.NewGuid(), _
     .Name = CType(attributes.Element(ns + "Publisher"), String) _
     }), _
     .BookAuthors = GetAuthors(ctx, attributes.Elements(ns + "Author")) _
     }

    Dim attachedSubjects As HashSet(Of Subject) = New HashSet(Of Subject)
    Dim attachedPublisher As HashSet(Of Publisher) = New HashSet(Of Publisher)

    For Each book As Book In booksToImport
      ctx.Books.InsertOnSubmit(book)

      If attachedSubjects.Contains(book.Subject) = False Then
        ctx.Subjects.Attach(book.Subject)
        attachedSubjects.Add(book.Subject)
      End If
      If attachedPublisher.Contains(book.Publisher) = False Then
        ctx.Publishers.Contains(book.Publisher)
        attachedPublisher.Add(book.Publisher)
      End If
    Next

    Try
      ctx.SubmitChanges()
      MessageBox.Show(booksToImport.Count().ToString() + " books imported.")
    Catch ex As Exception
      MessageBox.Show("An error occured while attempting to import the selected books. " + Environment.NewLine + Environment.NewLine + ex.Message)
    End Try
  End Sub

  Function GetAuthors(ByVal ctx As LinqInActionDataContext, ByVal authorElements As IEnumerable(Of XElement)) As EntitySet(Of BookAuthor)
    Dim bookAuthors = From authorElement In authorElements _
     Group Join author In ctx.Authors On CType(authorElement, String).Trim() Equals author.FirstName + " " + author.LastName Into Group _
     From existingAuthor In Group.DefaultIfEmpty() _
     Let nameParts = CType(authorElement, String).Split(" ") _
     Select New BookAuthor With { _
     .ID = Guid.NewGuid(), _
     .Author = If(existingAuthor IsNot Nothing, existingAuthor, New Author With { _
     .ID = Guid.NewGuid(), _
     .FirstName = nameParts(0), _
     .LastName = nameParts(1) _
     }) _
     }

    Dim authorSet As EntitySet(Of BookAuthor) = New EntitySet(Of BookAuthor)
    authorSet.AddRange(bookAuthors)
    Return authorSet
  End Function

  Function ParsePrice(ByVal priceElement As XElement) As Decimal
    Return Convert.ToDecimal(CType(priceElement.Element(ns + "FormattedPrice"), String).Replace("$", String.Empty))
  End Function
End Class
